x-pack/plugins/security_solution/server/lib/telemetry/sender.ts

```diff
@@ -0,0 +1,371 @@
+    const telemetryUrl = await this.telemetrySetup?.getTelemetryUrl();
+    if (!telemetryUrl) {
+      throw Error("Couldn't get telemetry URL");
+    }
+    return getV3UrlFromV2(telemetryUrl.toString(), 'alerts-endpoint'); // TODO: update
+  }
+
+  private async fetchLicenseInfo(): Promise<ESLicense | undefined> {
+    if (!this.core) {
+      return undefined;
+    }
+    try {
+      const callCluster = this.core.elasticsearch.legacy.client.callAsInternalUser;
+      const ret = await getLicense(callCluster, true);
+      return ret.license;
+    } catch (err) {
+      this.logger.warn(`Error retrieving license: ${err}`);
+      return undefined;
+    }
+  }
+
+  private copyLicenseFields(event: TelemetryEvent, lic: ESLicense | undefined) {
+    if (lic) {
+      event.license = {
+        uid: lic.uid,
+        status: lic.status,
+        type: lic.type,
+      };
+      if (lic.issued_to) {
+        event.license.issued_to = lic.issued_to;
+      }
+      if (lic.issuer) {
+        event.license.issuer = lic.issuer;
+      }
+    }
+  }
+
+  private async sendEvents(events: unknown[], telemetryUrl: string, clusterUuid: string) {
+    this.logger.debug(`Sending events: ${JSON.stringify(events, null, 2)}`);
+    const ndjson = transformDataToNdjson(events);
+    // this.logger.debug(`NDJSON: ${ndjson}`);
+
+    try {
+      const resp = await axios.post(`${telemetryUrl}?debug=true`, ndjson, {
+        // TODO: remove the debug
+        headers: {
+          'Content-Type': 'application/x-ndjson',
+          'X-Elastic-Cluster-ID': clusterUuid,
+          'X-Elastic-Telemetry': '1', // TODO: no longer needed?
+        },
+      });
+      this.logger.debug(`Events sent!. Response: ${resp.status} ${resp.data}`);
+    } catch (err) {
+      this.logger.warn(`Error sending events: ${err.response.status} ${err.response.data}`);
+    }
+  }
+}
+
+// For the Allowlist definition.
+interface AllowlistFields {
+  [key: string]: boolean | AllowlistFields;
+}
+
+// Allow list for the data we include in the events. True means that it is deep-cloned
+// blindly. Object contents means that we only copy the fields that appear explicitly in
+// the sub-object.
+const allowlistEventFields: AllowlistFields = {
+  '@timestamp': true,
+  agent: true,
+  Endpoint: true,
+  ecs: true,
+  elastic: true,
+  event: true,
+  file: {
+    name: true,
+    path: true,
+    size: true,
+    created: true,
+    accessed: true,
+    mtime: true,
+    directory: true,
+    hash: true,
+    Ext: {
+      code_signature: true,
+      malware_classification: true,
+    },
+  },
+  host: {
+    os: true,
+  },
+  process: {
+    name: true,
+    executable: true,
+    command_line: true,
+    hash: true,
+    Ext: {
+      code_signature: true,
+    },
+    parent: {
+      name: true,
+      executable: true,
+      command_line: true,
+      hash: true,
+      Ext: {
+        code_signature: true,
+      },
+    },
+  },
+};
+
+export function copyAllowlistedFields(
```

Comment:
nit -> I will do something like that, and I do not think that you need to cloneDeep here.

```
export function copyAllowlistedFields(
  allowlist: AllowlistFields,
  event: TelemetryEvent
): TelemetryEvent {
  return Object.entries(allowlist).reduce<TelemetryEvent>((newEvent, [allowKey, allowValue]) => {
    const eventValue = event[allowKey];
    if (eventValue) {
      if (allowValue === true) {
        return { ...newEvent, [allowKey]: eventValue };
      } else if (typeof allowValue === 'object' && typeof eventValue === 'object') {
        const values = copyAllowlistedFields(allowValue, eventValue as TelemetryEvent);
        return {
          ...newEvent,
          ...(Object.keys(values).length > 0 ? { [allowKey]: values } : {}),
        };
      }
    }
    return newEvent;
  }, {});
}
```